src/avif.c

```diff
@@ -184,39 +185,36 @@ avifResult avifImageCopy(avifImage * dstImage, const avifImage * srcImage, avifP
         if (allocationResult != AVIF_RESULT_OK) {
             return allocationResult;
         }
-
-        avifPixelFormatInfo formatInfo;
-        avifGetPixelFormatInfo(srcImage->yuvFormat, &formatInfo);
-        uint32_t uvHeight = (dstImage->height + formatInfo.chromaShiftY) >> formatInfo.chromaShiftY;
-        for (int yuvPlane = 0; yuvPlane < 3; ++yuvPlane) {
-            uint32_t planeHeight = (yuvPlane == AVIF_CHAN_Y) ? dstImage->height : uvHeight;
-
-            if (!srcImage->yuvRowBytes[yuvPlane]) {
+        for (int uvPlane = AVIF_CHAN_U; uvPlane <= AVIF_CHAN_V; ++uvPlane) {
+            if (!srcImage->yuvRowBytes[uvPlane]) {
                 // plane is absent. If we're copying from a source without
                 // them, mimic the source image's state by removing our copy.
-                avifFree(dstImage->yuvPlanes[yuvPlane]);
-                dstImage->yuvPlanes[yuvPlane] = NULL;
-                dstImage->yuvRowBytes[yuvPlane] = 0;
-                continue;
-            }
-
-            for (uint32_t j = 0; j < planeHeight; ++j) {
-                uint8_t * srcRow = &srcImage->yuvPlanes[yuvPlane][j * srcImage->yuvRowBytes[yuvPlane]];
-                uint8_t * dstRow = &dstImage->yuvPlanes[yuvPlane][j * dstImage->yuvRowBytes[yuvPlane]];
-                memcpy(dstRow, srcRow, dstImage->yuvRowBytes[yuvPlane]);
+                avifFree(dstImage->yuvPlanes[uvPlane]);
+                dstImage->yuvPlanes[uvPlane] = NULL;
+                dstImage->yuvRowBytes[uvPlane] = 0;
             }
         }
     }
-
     if ((planes & AVIF_PLANES_A) && srcImage->alphaPlane) {
         const avifResult allocationResult = avifImageAllocatePlanes(dstImage, AVIF_PLANES_A);
         if (allocationResult != AVIF_RESULT_OK) {
             return allocationResult;
         }
-        for (uint32_t j = 0; j < dstImage->height; ++j) {
-            uint8_t * srcAlphaRow = &srcImage->alphaPlane[j * srcImage->alphaRowBytes];
-            uint8_t * dstAlphaRow = &dstImage->alphaPlane[j * dstImage->alphaRowBytes];
-            memcpy(dstAlphaRow, srcAlphaRow, dstImage->alphaRowBytes);
+    }
+    for (int plane = AVIF_CHAN_Y; plane <= AVIF_CHAN_A; ++plane) {
+        uint32_t planeHeight = avifImagePlaneHeight(dstImage, plane);
+        uint8_t * srcRow = avifImagePlane(srcImage, plane);
+        uint8_t * dstRow = avifImagePlane(dstImage, plane);
+        if (!dstRow) {
+            continue;
+        }
```

Comment:
As I mentioned in https://github.com/AOMediaCodec/libavif/pull/1183#discussion_r1023834588, I believe this code is for `avifImage`s with `Y` and `U` not `NULL` and `V` `NULL`, for example. So checking the output of `avifImagePlaneHeight()` is not enough, because `avifImagePlaneHeight()` does not return `0` based on whether `yuvPlanes[V]` is `NULL` or not.

This way of handling bad input should be changed but this is beyond the scope of this PR.

Reply:
Thanks for the explanation.

In an earlier commit (https://github.com/AOMediaCodec/libavif/pull/1183/commits/68556c5e3740827b7ad50510bf63b3753511fad5) you were checking `srcRow` (presumably because I suggested it). I am just wondering why it's better to check `dstRow`. Both checks seem equivalent to me.

Reply:
I think the answer to my question is that the code allows a 4:0:0 `srcImage` to have non-null U and V planes, so we need to check `dstRow` instead.

Reply:
> I think the answer to my question is that the code allows a 4:0:0 `srcImage` to have non-null U and V planes, so we need to check `dstRow` instead.

Yes, and also the `avifPlanesFlags` argument may not be `AVIF_PLANES_ALL`.